Userland/Libraries/LibWeb/XHR/FormData.cpp

```diff
@@ -0,0 +1,154 @@
+ */
+
+#include <AK/TypeCasts.h>
+#include <LibJS/Runtime/Completion.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/FileAPI/Blob.h>
+#include <LibWeb/FileAPI/File.h>
+#include <LibWeb/HTML/FormControlInfrastructure.h>
+#include <LibWeb/WebIDL/DOMException.h>
+#include <LibWeb/XHR/FormData.h>
+
+namespace Web::XHR {
+
+// https://xhr.spec.whatwg.org/#dom-formdata
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, Optional<JS::NonnullGCPtr<HTML::HTMLFormElement>>)
+{
+    // FIXME: 1. If form is given, then:
+    // FIXME:    1. Let list be the result of constructing the entry list for form.
+    // FIXME:    2. If list is null, then throw an "InvalidStateError" DOMException.
+    // FIXME:    3. Set this’s entry list to list.
+
+    return construct_impl(realm);
+}
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list)
+{
+    return MUST_OR_THROW_OOM(realm.heap().allocate<FormData>(realm, realm, move(entry_list)));
+}
+
+FormData::FormData(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list)
+    : PlatformObject(realm)
+    , m_entry_list(move(entry_list))
+{
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::FormDataPrototype>(realm, "FormData"));
+}
+
+FormData::~FormData() = default;
+
+void FormData::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    for (auto entry : m_entry_list) {
+        for (auto value : entry.value) {
+            if (value.has<JS::Handle<FileAPI::File>>())
+                visitor.visit(value.get<JS::Handle<FileAPI::File>>());
+        }
+    }
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append
+WebIDL::ExceptionOr<void> FormData::append(DeprecatedString const& name, DeprecatedString const& value)
+{
+    return inner_append(name, value);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append-blob
+WebIDL::ExceptionOr<void> FormData::append(DeprecatedString const& name, JS::Handle<FileAPI::Blob> const& blob_value, DeprecatedString const& filename)
+{
+    return inner_append(name, blob_value, filename);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append
+// https://xhr.spec.whatwg.org/#dom-formdata-append-blob
+WebIDL::ExceptionOr<void> FormData::inner_append(DeprecatedString const& name, Variant<JS::Handle<FileAPI::Blob>, DeprecatedString> const& value, DeprecatedString const& filename)
+{
+    auto& realm = this->realm();
+
+    // 1. Let value be value if given; otherwise blobValue.
+    // 2. Let entry be the result of creating an entry with name, value, and filename if given.
+    auto entry = TRY(HTML::create_entry(realm, name, value, filename));
+
+    // 3. Append entry to this’s entry list.
+    if (auto entries = m_entry_list.get(entry.name); entries.has_value() && !entries->is_empty())
```

Comment:
I think you're copying `entries` here meaning this will not actually add an entry.